### ISSUE TITLE
Protractor 4 uncaughtException stop process

### ISSUE THREAD

**Issue by aluzardo** [protractor-cucumber-framework/protractor-cucumber-framework#56]
With Protractor 4 handling of errors thrown by Expected Conditions and Chai assertions (and possibly anything else) seems to have changed.

Now for example one `Not found element`exception stop and finish the process and do not continue with other tests.

Any possible fix in cucumber framework?

Issues in protractor
https://github.com/angular/protractor/issues/3491 https://github.com/angular/protractor/issues/3553

Issues in cucumberjs
https://github.com/cucumber/cucumber-js/issues/643

**Comment by bhreinb** [comment]
Hi there,

I'm seeing this issue too using v4.0.9 of protractor and is a bit of a blocker for cucumberjs users. See example stack trace below...

[12:30:36] E/launcher - Browser Hasn't Navigated To Expected Url 'http://172.16.87.184:54810/acaDev/#/welcome' Within Timeout '15000ms'
Wait timed out after 15001ms
[12:30:36] E/launcher - Error: Browser Hasn't Navigated To Expected Url 'http://172.16.87.184:54810/acaDev/#/welcome' Within Timeout '15000ms'
Wait timed out after 15001ms
    at C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:2364:22
    at ManagedPromise.invokeCallback_ (C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:1379:14)
    at TaskQueue.execute_ (C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:2913:14)
    at TaskQueue.executeNext_ (C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:2896:21)
    at C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:2775:27
    at C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:639:7
    at process._tickCallback (node.js:401:9)
From: Task: Browser Hasn't Navigated To Expected Url 'http://172.16.87.184:54810/acaDev/#/welcome' Within Timeout '15000ms'
    at ControlFlow.wait (C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:2352:17)
    at WebDriver.wait (C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\webdriver.js:712:29)
    at ProtractorBrowser.to.(anonymous function) [as wait] (C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\built\browser.js:62:29)
    at World.WaitTilWeNaviagateToFollowingUrlAndHaveAngular (C:\Users\bxb145\WebstormProjects\System_Test_Project\Support\World.js:146:19)
    at C:\Users\bxb145\WebstormProjects\System_Test_Project\ModuleHelper\Module_External_Login.js:364:40
    at TaskQueue.execute_ (C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:2913:14)
    at TaskQueue.executeNext_ (C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:2896:21)
    at C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:2775:27
    at C:\Users\bxb145\WebstormProjects\System_Test_Project\node_modules\protractor\node_modules\selenium-webdriver\lib\promise.js:639:7
    at process._tickCallback (node.js:401:9)
From: Task: <anonymous>
    at World.RunNonAngular (C:\Users\bxb145\WebstormProjects\System_Test_Project\Support\World.js:38:9)
    at AcaModuleLoginPageExternal.VerifyWeNavigateToUrlPatternAndCheckTitleOfPage (C:\Users\bxb145\WebstormProjects\System_Test_Project\ModuleHelper\Module_External_Login.js:361:22)
    at World.<anonymous> (C:\Users\bxb145\WebstormProjects\System_Test_Project\StepDefinitions\Background_steps.js:118:47)
    at nextTickCallbackWith0Args (node.js:452:9)
    at process._tickCallback (node.js:381:13)
[12:30:36] E/launcher - Process exited with error code 199

**Comment by NickTomlin** [comment]
Following up here from the protractor issue queue, the underlying problem here is that we don't wrap any of cucumber's step expectations to allow for resolving the webdriver control flow. It is surprising that things have just "worked" for so long.

To fix this properly, we'd need to ensure that cucumber's [step definitions](https://github.com/cucumber/cucumber-js/blob/f46fa68a29fd0ebfdc57e08a688d5ed8d2b6ab64/lib/cucumber/support_code/library.js#L73) are properly wrapped.

This will probably be similar to what Protactor does for [mocha](https://github.com/angular/protractor/blob/master/lib/frameworks/mocha.js#L24) or [jasmine](https://github.com/angular/jasminewd/blob/jasminewd2/index.js#L154).

**Comment by dancxjo** [comment]
+1

**Comment by dancxjo** [comment]
Here's my workaround:
```js
// step-wrapper.js
module.exports = function(scenario) {
    var originalDefineStep = scenario.defineStep;

    scenario.defineStep = function(pattern, fn) {
        originalDefineStep(pattern,
            function() {
                var args = arguments;
                browser.controlFlow().execute(() => fn.apply(this, args));
            }
        );

    };

    scenario.Then = scenario.defineStep;
    scenario.Given = scenario.defineStep;
    scenario.When = scenario.defineStep;
};

// From inside your step definitions
module.exports = function () {
   require('../../lib/step-wrapper')(this);
   ...
};
```

**Comment by AlexandrosD** [comment]
looks good. we'll need to try integrating this in the framework now we have
a solution.

**Comment by dancxjo** [comment]
Honestly, it's pretty hacky. I'd hope someone smarter than me could find a better solution. Unfortunately, Cucumber is smart about detecting the length of the callback function.

**Comment by AlexandrosD** [comment]
To me, it seems hacky, yes. But I doubt it is, since what needs to be done
to solve this issue and the relevant problems once and forever is to
enqueue cucumber's promises into the control flow, similarly to what is
done for jasmine. And your workaround does exactly this.

As a side note, this could've been easily sorted out if all the interacting
frameworks (read cucumber, protractor) would sit down and decide what needs
to be done, and how. That would be the most clear way of having this issue,
but...

**Comment by AlexandrosD** [comment]
Okay, I was trying to see where this could fit within the frameworks, experimented a bit with cucumber event handlers, but to no avail.

Only to find out that protractor v4.0.10 has added support for ignoring the uncaught exceptions. The only thing that's needed is to add the `ignoreUncaughtExceptions: true` property in the config file.

**Comment by darrinholst** [comment]
It sounds like this problem was "fixed" with the `ignoreUncaughtExceptions` option in protractor 4.0.10. I've been trying to reproduce this to add a test here, but I've been unsuccessful. Could somebody take a look at [this sample project](https://github.com/dholst/protractor-cucumber-framework-uncaught-exceptions) and tell me what I need to do to reproduce it?

It runs 3 scenarios [here](https://github.com/dholst/protractor-cucumber-framework-uncaught-exceptions/blob/master/features/uncaught.feature) with the step definition defined as [this](https://github.com/dholst/protractor-cucumber-framework-uncaught-exceptions/blob/master/features/step-definitions/defs.js#L2). It does throw the `NoSuchElementError` errors, but I see all 3 of them in the output...

```
✓ run test
(node:53775) DeprecationWarning: os.tmpDir() is deprecated. Use os.tmpdir() instead.
[12:43:00] I/local - Starting selenium standalone server...
[12:43:00] I/launcher - Running 1 instances of WebDriver
[12:43:01] I/local - Selenium standalone server started at http://192.168.0.19:58534/wd/hub
Feature: Uncaught

  Scenario:
  ✖ Given something throws an uncaught exception

  Scenario:
  ✖ Given something throws an uncaught exception

  Scenario:
  ✖ Given something throws an uncaught exception

Failures:

1) Scenario:  - features/uncaught.feature:2
   Step: Given something throws an uncaught exception - features/uncaught.feature:3
   Step Definition: features/step-definitions/defs.js:2
   Message:
     NoSuchElementError: No element found using locator: By(css selector, [nuthin])
         at /Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:719:27
         at ManagedPromise.invokeCallback_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:1379:14)
         at TaskQueue.execute_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:2913:14)
         at TaskQueue.executeNext_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:2896:21)
         at asyncRun (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:2775:27)
         at /Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:639:7
         at process._tickCallback (internal/process/next_tick.js:109:7)Error
         at ElementArrayFinder.applyAction_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:406:27)
         at ElementArrayFinder._this.(anonymous function) [as getText] (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:101:30)
         at ElementFinder.(anonymous function) [as getText] (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:742:22)
         at World.<anonymous> (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/features/step-definitions/defs.js:4:47)
         at _combinedTickCallback (internal/process/next_tick.js:73:7)

2) Scenario:  - features/uncaught.feature:5
   Step: Given something throws an uncaught exception - features/uncaught.feature:6
   Step Definition: features/step-definitions/defs.js:2
   Message:
     NoSuchElementError: No element found using locator: By(css selector, [nuthin])
         at /Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:719:27
         at ManagedPromise.invokeCallback_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:1379:14)
         at TaskQueue.execute_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:2913:14)
         at TaskQueue.executeNext_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:2896:21)
         at asyncRun (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:2775:27)
         at /Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:639:7
         at process._tickCallback (internal/process/next_tick.js:109:7)Error
         at ElementArrayFinder.applyAction_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:406:27)
         at ElementArrayFinder._this.(anonymous function) [as getText] (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:101:30)
         at ElementFinder.(anonymous function) [as getText] (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:742:22)
         at World.<anonymous> (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/features/step-definitions/defs.js:4:47)
         at _combinedTickCallback (internal/process/next_tick.js:73:7)

3) Scenario:  - features/uncaught.feature:8
   Step: Given something throws an uncaught exception - features/uncaught.feature:9
   Step Definition: features/step-definitions/defs.js:2
   Message:
     NoSuchElementError: No element found using locator: By(css selector, [nuthin])
         at /Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:719:27
         at ManagedPromise.invokeCallback_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:1379:14)
         at TaskQueue.execute_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:2913:14)
         at TaskQueue.executeNext_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:2896:21)
         at asyncRun (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:2775:27)
         at /Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/selenium-webdriver/lib/promise.js:639:7
         at process._tickCallback (internal/process/next_tick.js:109:7)Error
         at ElementArrayFinder.applyAction_ (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:406:27)
         at ElementArrayFinder._this.(anonymous function) [as getText] (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:101:30)
         at ElementFinder.(anonymous function) [as getText] (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/node_modules/protractor/built/element.js:742:22)
         at World.<anonymous> (/Users/darrinholst/Projects/protractor-cucumber-framework-uncaught-exceptions/features/step-definitions/defs.js:4:47)
         at _combinedTickCallback (internal/process/next_tick.js:73:7)

3 scenarios (3 failed)
3 steps (3 failed)
0m01.180s
[12:43:04] I/local - Shutting down selenium standalone server.
[12:43:04] I/launcher - 0 instance(s) of WebDriver still running
[12:43:04] I/launcher - chrome #01 failed 3 test(s)
[12:43:04] I/launcher - overall: 3 failed spec(s)
[12:43:04] E/launcher - Process exited with error code 1
~/Projects/protractor-cucumber-framework-uncaught-exceptions (master)
✗
```

versions are...

```
    "chai": "3.5.0",
    "chai-as-promised": "6.0.0",
    "cucumber": "1.3.2",
    "protractor": "4.0.9",
    "protractor-cucumber-framework": "3.0.0"
```

**Comment by darrinholst** [comment]
ok, I figured it out. I had to use the callback form since just returning the promise resulted in a rejected promise which I guess works fine. Anyway...instead of figuring out how to wrap both versions of cucumbers step definitions I think just documenting the `ignoreUncaughtExceptions` option is good enough for now. I did that [here](https://github.com/protractor-cucumber-framework/protractor-cucumber-framework#uncaught-exceptions).
